ja/documentation/ruby-from-other-languages/index.md

```diff
@@ -0,0 +1,441 @@
+* [PythonからRubyへ](/ja/documentation/ruby-from-other-languages/to-ruby-from-python/)
+
+## 重要な言語機能と知っておくべきこと
+
+ここでは、Rubyを学習する中で出会ういくつかの主要な機能について、
+参照先やヒントを示します。
+
+### イテレーション
+
+Rubyの特徴であるブロックとイテレータは、
+これまでにあまり見たことがない、あるいは使い慣れていないものかもしれません。
+(CやC++、1.5以前のJavaのような)インデックスを使った繰り返し処理や、
+(Perlの`for (@a) {...}`やPythonの`for i in aList: ...`のような)
+リストをループする処理の代わりに、
+Rubyでは次のようなコードをよく見ることになるでしょう。
+
+{% highlight ruby %}
+some_list.each do |this_item|
+  # ここはブロックの中になります。
+  # this_itemを扱う処理をここに書きます。
+end
+{% endhighlight %}
+
+`each`(や`collect`、`find`、`inject`、`sort`など)について詳しくは、
+`ri Enumerable`を参照してみてください（そして、次に
+`ri Enumerable#some_method`として具体的なメソッドを参照してみてください)。
+
+### すべては値
+
+式と文に違いはありません。
+すべては値を持ちます。`nil`であっても例外ではありません。
+Rubyでは次のように書くことが可能です。
+
+{% highlight ruby %}
+x = 10
+y = 11
+z = if x < y
+      true
+    else
+      false
+    end
+z # => true
+{% endhighlight %}
+
+### シンボルは軽量文字列ではない
+
+多くのRuby初学者はシンボルが何であるか、どう使えばよいのかを理解するのに苦戦します。
+
+シンボルは識別子として記述するのに適しています。
+シンボルは、**どんな物か**ではなく、**何者であるか**がすべてです。
+`irb`を立ち上げて、次の違いを見てみてください。
+
+{% highlight irb %}
+irb(main):001:0> :george.object_id == :george.object_id
+=> true
+irb(main):002:0> "george".object_id == "george".object_id
+=> false
+irb(main):003:0>
+{% endhighlight %}
+
+`object_id`メソッドはオブジェクトのIDを返します。
+もし二つのオブジェクトが同じ`object_id`を持っていれば、
+それらはメモリ上の同じ実体を指すものであると言えます。
+
+上記で見たとおり、一度でもシンボルを使ったなら、
+同じ文字からなるシンボルは、いずれもメモリ内の同じオブジェクトを参照します。
+同じ文字で表された2つのシンボルは、同じ`object_id`を指します。
+
+次に文字列("george")の方を見てみましょう。
+`object_id`は一致していません。
+これは、二つの文字列が異なる実体を指していることを示しています。
+新しい文字列を使うと、Rubyは毎回その文字列のために新しいメモリ領域を確保します。
+
+シンボルを使うか文字列を使うかで迷ったなら、
+ハッシュのキーのようにオブジェクトを識別したいのか、
+それとも前述の"george"などのようにコンテンツを表したいのかを、
+選択の指針にしてください。
+
+### すべてはオブジェクト
+
+「すべてはオブジェクト」は誇張ではありません。
+クラスや整数でさえオブジェクトです。
+他のオブジェクトと同じように扱うことができます。
+
+{% highlight ruby %}
+# 以下のクラス定義と同じことを行っています
+# class MyClass
+#   attr_accessor :instance_var
+# end
+MyClass = Class.new do
+  attr_accessor :instance_var
+end
+{% endhighlight %}
+
+### 可変な定数
+
+定数は厳密な意味での定数ではありません。
+初期化済みの定数を変更したとしても、
+警告はされますが、プログラムは停止しません。
+だからと言って、定数を再定義するべきだということではありません。
+
+### 名前付けの規約
+
+Rubyにはいくつか名前付けについての規約があります。
+大文字から始まる識別子は定数です。
+ドルマーク(`$`)から始まる識別子はグローバル変数、
+`@`から始まる識別子はインスタンス変数、
+`@@`から始まる識別子はクラス変数になります。
+
+メソッド名は大文字から始めることもできます。
+けれど、それは次のような混乱を招くことになるでしょう。
+
+{% highlight ruby %}
+Constant = 10
+def Constant
+  11
+end
+{% endhighlight %}
+
+こうすると、`Constant`は10ですが、`Constant()`は11になります。
+
+### キーワード引数
+
+メソッドはRuby2.0から、Pythonのように、
+キーワード引数を定義できるようになりました。
+
+{% highlight ruby %}
+def deliver(from: "A", to: nil, via: "mail")
+  "Sending from #{from} to #{to} via #{via}."
+end
+
+deliver(to: "B")
+# => "Sending from A to B via mail."
+deliver(via: "Pony Express", from: "B", to: "A")
+# => "Sending from B to A via Pony Express."
+{% endhighlight %}
+
+### 普遍の真理
+
+Rubyでは、`nil`と`false`を除くすべてのものは真と評価されます。
+CやPythonを始めとする多くの言語では、0や任意の正の値、空のリストなどは
+偽と評価されます。次に示すPythonコードをみてください
+(他の言語でもだいたい同じようなコードになるでしょう)。
+
+{% highlight python %}
+# Python版
+if 0:
+  print "0 is true"
```

Comment:
`print` should add `()` to fix syntax error in Python3. https://github.com/ruby/www.ruby-lang.org/pull/1374